### PR TITLE
Bank account and Authorized payer buttons displayed instead of Connect bank account

### PR DESCRIPTION
**Pull request by ikevin127**
### Details

Everytime we toggle "Make or track payments" we set the `achAccount: {reimburser: reimburserEmail}` and the UI ends up looking like we have a bank account added because of this [hasVBA](https://github.com/Expensify/App/blob/c50ae1790cabdb7ea0a0be872718d07b74c9ff4c/src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx#L87) check, when we don't actually have a bank account added yet.

To fix this we changed the `hasVBA` to check when we actually have a BA by checking for `bankAccountID`, additionally added the `isLoadingWorkspaceReimbursement` for the `SetWorkspaceReimbursement` call in order to show a loading spinner when we toggle "Make or track payments", instead of having the UI jump from the "Connect bank account" button to "Bank account" and "Authorized payer" buttons.

**Small refactor:** added `shouldShowBankAccount` variable because there were 3 occurances of this check. 

### Fixed Issues

$ https://github.com/Expensify/App/issues/39947
PROPOSAL: https://github.com/Expensify/App/issues/39947#issuecomment-2048010182

### Tests

1. Create a new workspace.
2. Go to **Workspace** > **More features** and enable **Workflows**
3. Go to **Workspace** > **Workflows** and under **Make or track payments** tap on **Connect bank account**.
4. Go back.
5. Disable **Make or track payments** then enable it again.
6. Verify that only the **Connect bank account** button is visible.

- [x] Verify that no errors appear in the JS console

### Offline tests

**TLDR:** same as **Tests**.

1. Create a new workspace.
2. Go to **Workspace** > **More features** and enable **Workflows**
3. Go to **Workspace** > **Workflows** and under **Make or track payments** tap on **Connect bank account**.
4. Go back.
5. Disable **Make or track payments** then enable it again.
6. Verify that only the **Connect bank account** button is visible.

### QA Steps

**TLDR:** same as **Tests**.

1. Create a new workspace.
2. Go to **Workspace** > **More features** and enable **Workflows**
3. Go to **Workspace** > **Workflows** and under **Make or track payments** tap on **Connect bank account**.
4. Go back.
5. Disable **Make or track payments** then enable it again.
6. Verify that only the **Connect bank account** button is visible.

**Additional tests from PR:**
- https://github.com/Expensify/App/pull/39017:

**Preconditions:** Having a policy with VBA and some Admins.

**Happy path:**
1. Open Settings > Workspaces > > Workflows.
2. See that you have an Authorized payer.
3. Click on the Authorized payer and change it to a different one.
4. Now, change it to one that has a display name set and check that the display name is shown under Authorized payer.

**Error path:**
1. Open Settings > Workspaces > > Workflows.
2. See that you have an Authorized payer.
3. Go offline and change the Authorized payer.
4. In a different tab/browser demote the Authorized payer from step 3 to Member.
5. Go back to the offline tab and get back online, the Authorized payer should get reverted to the previous one and there should be a red dot next to it indicating that it could not update.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/09330c1e-9988-4508-aaf2-e6650173d7fb

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/56457735/8513eb1d-ffd0-40e6-8e64-6598e622b276

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/6bcc4c5e-597e-49c5-8311-4e17d1f8ecc6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/56457735/3c2bdfe0-b799-4d91-a07a-3547b9fc7f85

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/81ee0b42-4d05-4dca-bc4b-0d8655cf3191

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/56457735/0d58cdf5-d10d-4cee-9b79-bf6d3d16b4c3

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
I think this looks pretty good, cc @Expensify/design (you too Tom) for extra eyes on the loading spinner pattern here.

**Comment by dubielzyk-expensify** [comment]
Looks good to me as well

**Comment by lakchote** [comment]
@ishpaul777 please wait to merge this PR as we need the input of the design team for the loading spinner ([context](https://github.com/Expensify/App/issues/39947#issuecomment-2055792768)).

**Comment by trjExpensify** [comment]
> I think this looks pretty good, cc @Expensify/design (you too Tom) for extra eyes on the loading spinner pattern here.

Yeah, I think the loading spinner is fine. It would be great if we could minimise that time to load, but this works for sure.

**Comment by ikevin127** [comment]
♻ Fixed conflict by removing unused type export `ACHAccount`, ready for review!

## 
> It would be great if we could minimise that time to load, but this works for sure.

@trjExpensify We can involve BE for this in a separate task regarding the `SetWorkspaceReimbursement` call, and I'd also make sure this bug that I found mentioned in https://github.com/Expensify/App/issues/39947#issuecomment-2052272735 is looked into by BE if they're going to try and minimise time to load for the `SetWorkspaceReimbursement ` call.

**Comment by ikevin127** [comment]
@ishpaul777 ✅

**Comment by ikevin127** [comment]
Indeed, ultimately the Workflow's no-BA / BA flow should be re-considered from a BE point of view and fixed properly on the FE side as all the proposals from this issue (including mine) are undesired workarounds (IMO) that don't really address the flow completely from the following points of view:
- seamless UI/UX
- happy path / unhappy path
- offline / online transiton (optimistically)
- BE response time regarding BA details (too slow currently)
- correct `SetWorkspaceReimbursement`call (currently it's not doing what it's supposed to do [set reimbursment], but doing extra things besides that), see [#39947 (comment)](https://github.com/Expensify/App/issues/39947#issuecomment-2052272735) bullet number 2 for more details)

cc @lakchote @ishpaul777

**Comment by ikevin127** [comment]
@trjExpensify If we want something solid here, it requires BE changes. Otherwise, I think FE only changes will only break the current functionality instead of actually improving the flow.

For more details see https://github.com/Expensify/App/pull/40182#issuecomment-2062656703. Additionally, I don't think any of the 2 options mentioned by https://github.com/Expensify/App/pull/40182#issuecomment-2067764395 are viable because we have BE related issues when it comes to the flow (mentioned in https://github.com/Expensify/App/pull/40182#issuecomment-2062656703).

**Comment by lakchote** [comment]
> Status for the PR: I feel the usage of loader here is not our pattern so i am not inclined to approve it, If anyone disagrees i'd love to hear a different POV, as of this comment from @ikevin127 i do agree there's a lot going on with command.

The loader pattern was initially approved by @Expensify/design team [here](https://github.com/Expensify/App/pull/40182#issuecomment-2054151153). 

@Expensify/design could you please check @ishpaul777's comment [here](https://github.com/Expensify/App/pull/40182#issuecomment-2058924420) regarding the infinite loader scenario?

**Comment by trjExpensify** [comment]
Why do we have to have an infinite loader when offline? Can't we do something like this?

image

**Comment by shawnborton** [comment]
Yeah, I like Tom's suggestion.

**Comment by dannymcclain** [comment]
I like Tom's suggestion too.

**Comment by ikevin127** [comment]
@trjExpensify @lakchote The infinite loader while offline behaviour was already fixed by commit https://github.com/Expensify/App/pull/40182/commits/61afb5bcf0c92e15c01555fc86bd8672b8ca39bb, and it matches the design suggested in https://github.com/Expensify/App/pull/40182#issuecomment-2102338268.

**Comment by ikevin127** [comment]
@ishpaul777 Just pushed the last commit 🤞 the PR is ready for review!

If there's anything unclear about the code changes please take a look at the updated [proposal](https://github.com/Expensify/App/issues/39947#issuecomment-2048010182), and if that doesn't answer your question, feel free to ask here!

**Comment by ikevin127** [comment]
@ishpaul777 In tried to reproduce, following your actions exactly like in the video for MacOS: Chrome and Android: Native and I was not able to reproduce the issue.

Must be related to API calls and probably because you are on emulator where things can be unpredictable network wise. I don't think this can be reproduced on a real device with a stable network connection as long as Expensify API is operational.

I also tried on both emulator and real device (OnePlus 7T - Android 14) and I could not reproduce, here are the videos:

<details>
<summary>Android: Native (Emulator / OnePlus 7T)</summary>

| Emulator | OnePlus 7T |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/a95a6690-3241-4aa1-bd67-209aa339fe78"> | <video src="https://github.com/Expensify/App/assets/56457735/da04ef35-7f74-483f-824b-da56616764b5"> |

</details>

> [!NOTE]
> I just synced with main.

@lakchote Can you help trigger a build for this PR so we can test this on real android devices ? This way we can hopefully move on with the review process!

**Comment by lakchote** [comment]
> @lakchote Can you help trigger a build for this PR so we can test this on real android devices ? This way we can hopefully move on with the review process!
>

Good idea, I've triggered a build and will let you know once it's ready.

**Comment by shawnborton** [comment]
Nice - are there updated screenshots for us to review?

**Comment by shawnborton** [comment]
In this screen, why does one row appear to be more faded out than the other?
CleanShot 2024-05-17 at 09 04 19@2x

Also, the two rows feel like they are too far apart. Is that something we can adjust here too?

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx [hunk 1]
```diff
@@ -168,60 +171,63 @@ function WorkspaceWorkflowsPage({policy, betas, route}: WorkspaceWorkflowsPagePr
...
-                        <MenuItem
-                            titleStyle={hasVBA ? styles.textLabelSupportingNormal : styles.textLabelSupportingEmptyValue}
-                            descriptionTextStyle={styles.textNormalThemeText}
-                            title={
-                                hasVBA && policy?.reimbursementChoice === CONST.POLICY.REIMBURSEMENT_CHOICES.REIMBURSEMENT_YES
-                                    ? translate('common.bankAccount')
-                                    : translate('workflowsPage.connectBankAccount')
-                            }
-                            description={bankDisplayName}
-                            disabled={isOffline || !isPolicyAdmin}
```
[the remark this answers is not in the document]
Comment (shawnborton) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:191]: Okay, I think we should not have it double-faded so let's definitely solve that

**Comment by shawnborton** [comment]
Can we try the second menu item with no top margin? This way it's 12px + 12px away from the item above it.

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:191]: > [!note]
> 1. The first disabled grayed-out fade comes from the "**Make or track payments**" toggle when toggled while offline.
> 2. The second one comes from **Connect bank account / Bank account** option which was and  is disabled while offline.
> 3. **Authorized payer** option is enabled while offline -> meaning the **Authorized payer** can be changed optimistically. This is why this one doesn't get double disabled grayed-out fade when "**Make or track payments**" is toggled while offline. 

**ishpaul777**:
> we can add a check !policy?.pendingFields?.reimbursementChoice  && isOffline to solve this

I don't think this logic is good since it would break the logic mentioned at (2.) where **Connect bank account / Bank account** option was and is supposed to be disabled while offline.

**shawnborton**:
> Okay, I think we should not have it double-faded so let's definitely solve that

To do this, instead I would use **Connect bank account / Bank account** option `shouldGreyOutWhenDisabled` to not have the 2nd gray-out fade applied to the option when already applied from "**Make or track payments**" toggle. 

Similarly, we have to do the same for the `OfflineWithFeedback` wrapper of **Authorized payer** option (3.) we add:
```ts
shouldDisableOpacity={isOffline && policy?.pendingFields?.reimbursementChoice && policy?.pendingFields?.reimburser}
```
in order to not double gray-out the **Authorized payer** if already applied from "**Make or track payments**" toggle, when **Authorized payer** is changed while offline. This is how it would look in all scenarios:

1. **Without Bank account added:**

| Offline (can't add / change **Connect bank account** option) |
| ------------- |
| noba-offline |

| Offline (with "Make or track payments" Toggled (optimistically)) |
| ------------- |
| noba-offline-toggled |

2. **With Bank account added:**

| Offline (can't add / change **Bank account** option) |
| ------------- |
| ba-offline |

| Offline (with "Make or track payments" Toggled (optimistically)) |
| ------------- |
| ba-offline-toggled-1 |

| Offline (with "Make or track payments" Toggled + Authorized payer changed (optimistically)) |
| ------------- |
| ba-offline-toggled-2 |

| Offline (with "Make or track payments" Toggled + Authorized payer changed (optimistically) and hover on Authorized payer) |
| ------------- |
| ba-offline-toggled-2-hover |

```ts
shouldDisableOpacity={isOffline && policy?.pendingFields?.reimbursementChoice && policy?.pendingFields?.reimburser}
```
^ - Writing it like this is specifically important for the below case, since we don't want to NOT gray out **Authorized payer** when changed offline (optimistically) in case "Make or track payments" was NOT toggled, by only writing it like:
❌ `shouldDisableOpacity={policy?.pendingFields?.reimburser}`
since that would NOT gray out the optimistically set **Authorized payer**, instead we want to gray it out, maintaining the pattern like so:

| Offline (with "Make or track payments" Not Toggled + Authorized payer changed (optimistically)) |
| ------------- |
| ba-offline-non-toggled-changed-payer |

Please let me know if we're in agreement with the changes mentioned above from a design / code POV.
Once agreed I'll push the changes!

cc @ishpaul777 @shawnborton

**Comment by ikevin127** [comment]
> Can we try the second menu item with no top margin? This way it's 12px + 12px away from the item above it.

Did this for second menu item, this are the results:

| Before | After |
| ------------- | ------------- |
| 2nd-mt4 | 2nd-mt0 |

Please let me know if we're in agreement with the changes mentioned above. Once agreed I'll push the changes!

cc @shawnborton

**Comment by shawnborton** [comment]
> Did this for second menu item, this are the results:

Looks great!

**Comment by ikevin127** [comment]
@ishpaul777 Just pushed changes from [here](https://github.com/Expensify/App/pull/40182#discussion_r1605594216) and [here](https://github.com/Expensify/App/pull/40182#issuecomment-2118462210). PR is ready for review!

**Comment by ikevin127** [comment]
@ishpaul777 Just pushed changes from [here](https://github.com/Expensify/App/pull/40182#discussion_r1605594216) and [here](https://github.com/Expensify/App/pull/40182#issuecomment-2118462210). PR is ready for review!

**Review by ishpaul777 — APPROVED** [review]
LGTM!

**Review by lakchote — DISMISSED** [review]

**Comment by ikevin127** [comment]
@lakchote Are we merging this soon ? I synced w/ main because it had conflicts.

**Comment by lakchote** [comment]
> @lakchote Are we merging this soon ? I synced w/ main because it had conflicts.

There was a merge freeze. Now that it's been removed, I'm going to merge this.

**Review by lakchote — APPROVED** [review]
